pkg/authorization/rules_engine.go

```diff
@@ -0,0 +1,37 @@
+package authorization
+
+import (
+	"net/http"
+
+	"github.com/oauth2-proxy/oauth2-proxy/pkg/apis/sessions"
+)
+
+// engine manages authorization Rules and which requests
+// pass or fail authorization
+type engine struct {
+	rules         []Rule
+	defaultPolicy Policy
+}
+
+// NewRulesEngine builds a RulesEngine for HTTP request based authZ
+func NewRulesEngine(rules []Rule, defaultPolicy Policy) RulesEngine {
+	return &engine{
+		rules:         rules,
+		defaultPolicy: defaultPolicy,
+	}
+}
```

Comment:
Why not construct the rules from the options in here? This would simplify the initialisation within the OAuth2 Proxy construction?